### PR TITLE
[Music] Gracefully handle changes to level artist packs

### PR DESCRIPTION
**Pull request by mikeharv**
Yesterday, @onlinecsteacher hit upon an issue where sound blocks had unexpected values after changing the selected artist pack in a level. She said:

> I changed the song pack in this [level](https://levelbuilder-studio.code.org/s/music-lab-advanced-model-sandbox/lessons/2/levels/1) from Espresso to Chantaje but all of the sound options are still from the Espresso pack
> If you change the song pack and don’t update the sound blocks, they’ll keep the original sounds and not default to the first drum/bass beat of the new pack
> I fixed it on this level by switching back to intro2024 and then launch 2024

|**Before changing packs**|**After changing packs**|
|-|-|
|image|image|

We recently added a field validator to ensure that a valid value is always set in sound blocks. https://github.com/code-dot-org/code-dot-org/pull/61727 However, just because a sound is found in the library, doesn't mean it is part of the "allowed" set for the level. This branch updates the validator to also check whether a sound is allowed.

After implementing this, I noticed that it was possible to get actually get an undefined default sound. The `getAvailableSounds()` function filters the currently allowed sounds using the current pack id. Both the allowed sounds and pack id are part of the level data, and it's actually possible for them to conflict if the pack is the last setting to be toggled.

image (236)

The reason Emma's workaround works is because we reset all sound options when a new library is chosen.

This branch introduces a second change which is to (partially) reset the sound options when an artist pack is chosen. However, rather than wiping all sound options, what we really care about are the library's restricted packs (ie. our licensed songs). Rather than simply updating the level data with an undefined sounds value, I opted to preserve any selected sounds from our unrestricted/generic sound packs.

https://github.com/user-attachments/assets/56179b0b-da1e-451f-9620-096388ccc94c

Together, these changes should help ensure that the sound options are always aligned with the pack id, if either values are present. They will also better ensure that sound fields always have a valid and allowed sound value.

Another possibility that has been considered, is that we could reset or modify the start sources JSON whenever libraries, artist packs, or sound options are toggled. Removing all start sources feels very destructive, and surgically modifying the start sources field values wouldn't have any benefit over relying on the field validator.

In a perfect world, the levelbuilder would always configure these options before setting start sources. However, in practice, there's a fair bit more iteration involved in crafting a perfect level. (The only change that forcefully resets start sources is the block mode. Each block mode uses totally separate block definitions, so the sources cannot be compatible with both.)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lab2/levelEditors/levelData/EditMusicLevelData.tsx [hunk 1]
```diff
@@ -84,6 +84,8 @@ const EditMusicLevelData: React.FunctionComponent<EditMusicLevelDataProps> = ({
     [levelData.library, loadedLibraries]
   );
 
+  const packKeys = (restrictedPacks || []).map(pack => pack.value) || [];
```
Comment (sanchitmalhotra126) [inline apps/src/lab2/levelEditors/levelData/EditMusicLevelData.tsx]: nit: maybe `restrictedPackKeys` for clarity?

**Review by sanchitmalhotra126 — APPROVED** [review]
Nice updates!

File: apps/src/music/blockly/extensions.js [hunk 2]
```diff
@@ -155,12 +155,33 @@ export const effectsFieldExtension = function () {
   };
 };
 
+/**
+ * Extension to blocks with sound fields that validates new values.
+ */
 export const fieldSoundsValidator = function () {
+  /**
+   * Ensures that sound blocks also have a valid value, even if a level's library or song
+   * pack has changed. *
```
Comment (breville) [inline apps/src/music/blockly/extensions.js]: nit: stray `*`?

**Comment by breville** [comment]
> hit upon an issue where sound blocks had unexpected values after changing the selected artist pack in a level

To clarify, this happened in levelbuilder?  Or in a student-facing level when the artist pack was changed?

**Comment by mikeharv** [comment]
> > hit upon an issue where sound blocks had unexpected values after changing the selected artist pack in a level
> 
> To clarify, this happened in levelbuilder? Or in a student-facing level when the artist pack was changed?

This happened on levelbuilder, but the incident revealed it could happen in a student-facing level if the start sources are incompatible with the artist pack or sound options.

**Pull request merged**